agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,563 @@ package gateways
+		_, boundRefs, bindErrors := BindRoutesToGateways([]*gatewayMeta{meta}, route)
+		// unset the old gateway binding in case it's stale
+		for _, parent := range route.GetParents() {
+			if parent.Kind == gateway.Kind && parent.Name == parent.Name && parent.EnterpriseMeta.IsSame(&gateway.EnterpriseMeta) {
+				routeUpdater.RemoveCondition(structs.Condition{
+					Type:     "Bound",
+					Resource: &parent,
+				})
+			}
+		}
+		for _, ref := range boundRefs {
+			routeUpdater.SetCondition(structs.Condition{
+				Type:               "Bound",
+				Status:             "True",
+				Reason:             "Bound",
+				Resource:           &ref,
+				Message:            "successfully bound route",
+				LastTransitionTime: &now,
+			})
+		}
+		for reference, err := range bindErrors {
+			routeUpdater.SetCondition(structs.Condition{
+				Type:               "Bound",
+				Status:             "False",
+				Reason:             "FailedToBind",
+				Resource:           &reference,
+				Message:            err.Error(),
+				LastTransitionTime: &now,
+			})
+		}
+		if entry, updated := routeUpdater.UpdateEntry(); updated {
+			updatedRoutes = append(updatedRoutes, entry)
+		}
+	}
+
+	// first check for gateway conflicts
+	for i, listener := range meta.BoundGateway.Listeners {
+		protocol := meta.Gateway.Listeners[i].Protocol
+		switch protocol {
+		case structs.ListenerProtocolTCP:
+			if len(listener.Routes) > 1 {
+				updater.SetCondition(structs.Condition{
+					Type:    "Conflicted",
+					Status:  "True",
+					Reason:  "RouteConflict",
+					Message: "TCP-based listeners currently only support binding a single route",
+					Resource: &structs.ResourceReference{
+						Kind:           structs.APIGateway,
+						Name:           meta.Gateway.Name,
+						SectionName:    listener.Name,
+						EnterpriseMeta: meta.Gateway.EnterpriseMeta,
+					},
+					LastTransitionTime: &now,
+				})
+				continue
+			}
+		}
+		updater.SetCondition(structs.Condition{
+			Type:   "Conflicted",
+			Status: "False",
+			Reason: "NoConflict",
+			Resource: &structs.ResourceReference{
+				Kind:           structs.APIGateway,
+				Name:           meta.Gateway.Name,
+				SectionName:    listener.Name,
+				EnterpriseMeta: meta.Gateway.EnterpriseMeta,
+			},
+			Message:            "listener has no route conflicts",
+			LastTransitionTime: &now,
+		})
 	}
 
-	//initialize object, values get copied over in ensureBoundGateway if they don't exist
-	metaGateway.BoundGateway = boundGateway.(*structs.BoundAPIGatewayConfigEntry)
-	return metaGateway, nil
+	// now check if we need to update the gateway status
+	if toUpdate, shouldUpdate := updater.UpdateEntry(); shouldUpdate {
+		r.logger.Debug("persisting gateway status", "gateway", gateway)
+		if err := r.updater.UpdateWithStatus(toUpdate); err != nil {
+			r.logger.Error("error persisting gateway status", "error", err)
+			return err
+		}
+	}
+
+	// next update route statuses
+	for _, toUpdate := range updatedRoutes {
```

Comment:
We only ever return routes that are newly bound/unbound from the gateway, so this should effectively act like the dirty checking mentioned above.